seacatauth/authn/webauthn/service.py

```diff
@@ -0,0 +1,422 @@
+import base64
+import datetime
+import json
+import logging
+import secrets
+import urllib.parse
+import re
+
+import asab.storage
+import webauthn
+import webauthn.registration
+import webauthn.helpers.structs
+
+#
+
+L = logging.getLogger(__name__)
+
+#
+
+
+class WebAuthnService(asab.Service):
+	WebAuthnCredentialCollection = "wa"
+	WebAuthnRegistrationChallengeCollection = "warc"
+
+	def __init__(self, app, service_name="seacatauth.WebAuthnService"):
+		super().__init__(app, service_name)
+		self.StorageService = app.get_service("asab.StorageService")
+		self.CredentialsService = app.get_service("seacatauth.CredentialsService")
+
+		self.RelyingPartyName = asab.Config.get("seacatauth:webauthn", "relying_party_name")
+
+		self.Origin = asab.Config.get("seacatauth:webauthn", "origin", fallback=None)
+		if self.Origin is None:
+			auth_webui_base_url = asab.Config.get("general", "auth_webui_base_url")
+			parsed = urllib.parse.urlparse(auth_webui_base_url)
+			self.Origin = "{}://{}".format(parsed.scheme, parsed.netloc)
+
+		# RP ID must match host's domain name (without scheme, port or subpath)
+		# https://www.w3.org/TR/webauthn-2/#relying-party-identifier
+		self.RelyingPartyId = asab.Config.get("seacatauth:webauthn", "relying_party_id", fallback=None)
+		if self.RelyingPartyId is None:
+			self.RelyingPartyId = str(urllib.parse.urlparse(self.Origin).hostname)
+
+		self.RegistrationTimeout = asab.Config.getseconds("seacatauth:webauthn", "challenge_timeout") * 1000
+		self.SupportedAlgorithms = [
+			webauthn.helpers.structs.COSEAlgorithmIdentifier(-7)  # Es256
+		]
+
+		self.KeyNameRegex = re.compile(r"^[a-z][a-z0-9._-]{0,128}[a-z0-9]$")
+
+		self.App.PubSub.subscribe("Application.tick/10!", self._on_tick)
+
+
+	async def _on_tick(self, event_name):
+		await self.delete_expired_challenges()
+
+
+	async def create_webauthn_credential(
+		self,
+		credentials_id: str,
+		verified_registration,  # webauthn.registration.VerifiedRegistration
+		name: str = None
+	):
+		"""
+		Create database entry for a verified WebAuthn credential
+		"""
+		if name is None:
+			name = "key-{}".format(datetime.datetime.now().strftime("%y%m%d-%H%M%S"))
```

Comment:
Is it key?
That's not FIDO2/Webauthn terminology (and it is visible at UI).

Reply:
This name should be only informative to the user and preferably unique. It is not used for identification anywhere. The user can also change this name to whatever they like.

The main reason I chose `key` instead of `credential` is that I found the word more user friendly, although I'm aware that it is not technically correct :grin: 

Should I change the default name to `credential-{datetime}` instead? Or do you have a better suggestion?

Reply:
I also considered the `credential-{counter}` format, but this would require a persistent per-user counter and I didn't want to clutter the SCA credentials object even more.

Reply:
After discussion I'm keeping `key-{date}-{time}`.